Add copy/paste feature to memory view

Vancleeff opened the pull request:
Enhanced user experience by providing a way to copy and paste value directly in the memory view.

Vancleeff committed:
Add copy/paste feature to memory view
Enhanced user experience by providing a way to copy and paste value directly in the memory view.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -231,6 +232,56 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                auto nAddress = m_pViewModel.GetAddress();
+                const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+                std::wstring sClipboardText = ra::services::ServiceLocator::Get<ra::services::IClipboard>().GetText();
+                auto n = m_pViewModel.GetSize();
```
`n` is not a very descriptive variable name, and I can't tell if it's even being used anywhere.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -231,6 +232,56 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+            if (bControlHeld)
+            {
+                const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+                const auto iValue = pEmulatorContext.ReadMemory(m_pViewModel.GetAddress(), m_pViewModel.GetSize());
```
please use `nValue` for consistency with the rest of the code.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -231,6 +232,56 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+            return true;
+
+        case 'V':
+            if (bControlHeld)
```
Paste shouldn't be allowed in the pointer finder. You'll have to expose `MemoryViewerViewModel::m_bReadOnly` and check that here.

Vancleeff committed:
Refactor copy/paste feature following review

 - New HandleShortcut function implemented to ViewModel
 - Improve code to match repository naming convention
 - Prevent from using paste feature in the pointer finder
 - Using of ParseHex instead of iswxdigit and refactoring logic based on this change

vancleeff commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -231,6 +232,56 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                for (wchar_t ch : sClipboardText)
+                    if (!iswxdigit(ch)) return false;
+
+                // Padding zeroes depending if shift is pressed (strict mode) or not (replace mode)
```
The padding was made for the shift mode : in shift mode, the user was able to paste a long hexadecimal sequence and it was pasted and replaced value in memory.
Example : 12345 pasted into these words "1234 0000" resulted in "1234 5000"
 
I changed that in the last commit to be more strict to avoid confusion. 
The shift mode still let the user to paste long sequence but if they are not a multiple of the current size, the residual part will be converted without padding zeroes :
Example : 12345 pasted into these words "1234 0000" result now in "1234 0005"

vancleeff commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -231,6 +232,56 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                }else 
+                    sClipboardText = sClipboardText.length() % 2 == 1 ? (L"0" + sClipboardText) : sClipboardText;
+
+                // Writing every byte separately considerably improves stability and enables long sequences to be pasted
```
It was unstable with big nword like "0x80123456" but with the implementation of ParseHex as you suggested, everything is now perfectly fine, thanks ! It was probably due to something wrong with the conversion of my wstring previously.

vancleeff commented:
Thanks for the review @Jamiras !
I commited these improvements !

Jamiras commented on src/ui/win32/bindings/ControlBinding.hh:
```diff
@@ -5,6 +5,8 @@
 #include "ui\BindingBase.hh"
 #include "ui\win32\DialogBase.hh"
 
+#include "ui\viewmodels\MessageBoxViewModel.hh"
```
Why was this moved? Please keep unnecessary headers out of header files as they have to be parsed for every file that includes the container header file.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -236,6 +239,68 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+                std::wstring sClipboardText = ra::services::ServiceLocator::Get<ra::services::IClipboard>().GetText();
+                const auto nNibblesForSize = ra::data::MemSizeBytes(m_pViewModel.GetSize());
+                std::vector<std::uint32_t> v_nValues;
```
The "x_" prefix defines the scope, and is only for member variables `m_`, statics `s_`, or globals `g_`. This should simply be named `vValues`.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -236,6 +239,68 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                auto nAddress = m_pViewModel.GetAddress();
+                const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+                std::wstring sClipboardText = ra::services::ServiceLocator::Get<ra::services::IClipboard>().GetText();
+                const auto nNibblesForSize = ra::data::MemSizeBytes(m_pViewModel.GetSize());
```
This should be `nBytesForSize`

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -236,6 +239,68 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                {
+                    //Single mode writes only the first value, multi mode (shift) writes every values
+                    pEmulatorContext.WriteMemory(nAddress, m_pViewModel.GetSize(), nValue);
+                    if (!bShiftHeld)
```
Recommend moving the shift handling logic into the above loop. No reason to populate the entire array if you're only going to process a single item.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -236,6 +239,68 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+
+        case 'V':
+            if (bControlHeld and !m_pViewModel.IsReadOnly())
+            {
```
This logic is sizable enough that I'd split it out into it's own function `OnPaste(bShiftHeld)`. Suggest doing the same for copy as well for consistency.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -236,6 +239,68 @@ bool MemoryViewerControlBinding::HandleNavigation(UINT nChar)
+                if (sClipboardText.empty())
+                    return false;
+
+                // Split the clipboard value into substrings matching the current size and check if they're valid hexadecimal values
```
Recommend documenting the shift+paste functionality, as it's something that is not normal and won't be obvious when looking at this code some time in the future.
```
// If the clipboard text is smaller than a single write, it will be treated as a padded value:
//    C => 0C, 000C, or 0000000C (depending on currently selected viewer size)
// If shift is held, and the clipboard string is bigger than a single write, multiple writes will occur. 
// If the last write is not a complete chunk, it will be padded in the same way as described above:
//    12345678C => 12 34 56 78 0C, 1234 5678 000C, or 12345678 0000000C
```

Vancleeff committed:
Move copy/paste logic to their own functions OnCopy/OnPaste

vancleeff commented:
Thanks @Jamiras, commited !

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -290,6 +319,66 @@ void MemoryViewerControlBinding::OnViewModelIntValueChanged(const IntModelProper
+    const auto nValue = pEmulatorContext.ReadMemory(m_pViewModel.GetAddress(), m_pViewModel.GetSize());
+    std::wstring sValue = ra::data::MemSizeFormat(nValue, m_pViewModel.GetSize(), MemFormat::Hex);
+
+    ra::services::ServiceLocator::Get<ra::services::IClipboard>().SetText(ra::Widen(sValue));
```
`ra::Widen` is unnecessary. `sValue` is already a `std::wstring`.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -290,6 +319,66 @@ void MemoryViewerControlBinding::OnViewModelIntValueChanged(const IntModelProper
+    {
+        pEmulatorContext.WriteMemory(nAddress, m_pViewModel.GetSize(), nValue);
+        if (!bShiftHeld)
+            break;
```
This break isn't needed. Processing will be aborted on line 364.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -290,6 +319,66 @@ void MemoryViewerControlBinding::OnViewModelIntValueChanged(const IntModelProper
+            return true;
+        }
+
+        vValues.push_back(nValue);
```
If you don't care about doing a full validation pass before writing, you can avoid using an array entirely and just write as you parse:
```
pEmulatorContext.WriteMemory(nAddress, m_pViewModel.GetSize(), nValue);
if (!bShiftHeld)
    return true;
nAddress += nBytesForSize;
```

Vancleeff committed:
Remove pointless array logic in OnPaste() function

Vancleeff committed:
Fix CI / Win64-Analysis checks

Vancleeff commented:
Fix CI checks

Jamiras commented:
This has conflicts from the +/- PR.
